src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/DefaultRazorCompletionFactsService.cs

```diff
@@ -4,97 +4,42 @@
-        {
-            CSharpCodeParser.AddTagHelperDirectiveDescriptor,
-            CSharpCodeParser.RemoveTagHelperDirectiveDescriptor,
-            CSharpCodeParser.TagHelperPrefixDirectiveDescriptor,
-        };
-
-        public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorCodeDocument codeDocument, SourceSpan location)
-        {
-            if (codeDocument is null)
-            {
-                throw new ArgumentNullException(nameof(codeDocument));
-            }
-
-            var completionItems = new List<RazorCompletionItem>();
-            var syntaxTree = codeDocument.GetSyntaxTree();
-
-            if (AtDirectiveCompletionPoint(syntaxTree, location))
-            {
-                var directiveCompletions = GetDirectiveCompletionItems(syntaxTree);
-                completionItems.AddRange(directiveCompletions);
-            }
+        private readonly IEnumerable<RazorCompletionItemProvider> _completionItemProviders;
 
-            return completionItems;
-        }
-
-        // Internal for testing
-        internal static List<RazorCompletionItem> GetDirectiveCompletionItems(RazorSyntaxTree syntaxTree)
+        [ImportingConstructor]
+        public DefaultRazorCompletionFactsService([ImportMany] IEnumerable<RazorCompletionItemProvider> completionItemProviders)
         {
-            var defaultDirectives = FileKinds.IsComponent(syntaxTree.Options.FileKind) ? Array.Empty<DirectiveDescriptor>() : DefaultDirectives;
-            var directives = syntaxTree.Options.Directives.Concat(defaultDirectives);
-            var completionItems = new List<RazorCompletionItem>();
-            foreach (var directive in directives)
+            if (completionItemProviders is null)
             {
-                var completionDisplayText = directive.DisplayName ?? directive.Directive;
-                var completionItem = new RazorCompletionItem(
-                    completionDisplayText,
-                    directive.Directive,
-                    directive.Description,
-                    RazorCompletionItemKind.Directive);
-                completionItems.Add(completionItem);
+                throw new ArgumentNullException(nameof(completionItemProviders));
             }
 
-            return completionItems;
+            _completionItemProviders = completionItemProviders;
         }
 
-        // Internal for testing
-        internal static bool AtDirectiveCompletionPoint(RazorSyntaxTree syntaxTree, SourceSpan location)
+        public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorCodeDocument codeDocument, SourceSpan location)
         {
-            if (syntaxTree == null)
-            {
-                return false;
-            }
-
-            var change = new SourceChange(location, string.Empty);
-            var owner = syntaxTree.Root.LocateOwner(change);
-
-            if (owner == null)
+            if (codeDocument is null)
```

Comment:
gross

Reply:
Ya ctrl Dot does this flavor of null checking by default now. Figured it wasn't worth fighting the tooling.

Reply:
I'm sad about this, but ok